pyface/ui/qt4/data_view/data_view_item_model.py

```diff
@@ -0,0 +1,235 @@
+# (C) Copyright 2005-2020 Enthought, Inc., Austin, TX
+# All rights reserved.
+#
+# This software is provided without warranty under the terms of the BSD
+# license included in LICENSE.txt and may be redistributed only under
+# the conditions described in the aforementioned license. The license
+# is also available online at http://www.enthought.com/licenses/BSD.txt
+#
+# Thanks for using Enthought open source!
+
+import logging
+
+from pyface.qt import is_qt5
+from pyface.qt.QtCore import QAbstractItemModel, QModelIndex, Qt
+from pyface.data_view.index_manager import Root
+from pyface.data_view.abstract_data_model import AbstractDataModel
+
+
+logger = logging.getLogger(__name__)
+
+# XXX This file is scaffolding and may need to be rewritten
+
+
+class DataViewItemModel(QAbstractItemModel):
+    """ A QAbstractItemModel that understands AbstractDataModels. """
+
+    def __init__(self, model, parent=None):
+        super().__init__(parent)
+        self.model = model
+        self.showRowHeader = True
+
+    @property
+    def model(self):
+        return self._model
+
+    @model.setter
+    def model(self, model: AbstractDataModel):
+        if hasattr(self, '_model'):
+            # disconnect trait listeners
+            self._model.observe(
+                self.on_structure_changed,
+                'structure_changed',
+                remove=True,
+            )
+            self._model.observe(
+                self.on_values_changed,
+                'values_changed',
+                remove=True,
+            )
```

Comment:
This isn't a widget, this is a `QAbstractItemModel` subclass (ie, a `QObject` subclass).  There are no add/remove event listener methods, nor is there a `dispose` method.  I don't think this is a worry in this case.

Reply:
I mean the `DataViewWidget` might not get garbage collected, and this `DataViewItemModel` (being the model of `DataViewWidget`) might still be alive, and then these change handlers would also be alive and responding to changes on the model.  This comment kind of spans two files... this is the related comment: https://github.com/enthought/pyface/pull/543/files#r449468439

Reply:
> I mean the DataViewWidget might not get garbage collected

Sorry again...I meant the `control` of `DataViewWidget` that might not be garbage collected.

Reply:
No, this is not an issue.  `DataViewItemModel` can potentially be _shared_ between views.  We _don't_ want `DataViewItemModel` to go, we just want to drop it from the `DataViewWidget` (which we now do in that class by setting the model to `None`).  It can happily hang around for as long as it needs to, listening to trait events and emitting change signals to anyone who cares to listen (or no-one).

It's not an explicit design goal right now, but it should be possible for someone to write an alternative widget which uses this class unmodified (eg. something which uses a `QTableView` instead of a `QTreeView`), possibly sharing this between multiple views (eg. linked tables that scroll together).  Trying to "clean up" inappropriately would make this sort of thing impossible.

This is the correct way to handle this, as far as I can see.

Reply:
I see. I don't know how that sharing scenario looks like. But actually `observe` is not idempotent. It is like matching parentheses: Each `observe(..., remove=False)` is matched with one `observe(..., remove=True)`. 
 This means two views can each call `observe(..., remove=False)` once, and when one of the views is done with it, it can call `observe(..., remove=True)` without affecting the second view.

`on_trait_change` would not have allowed for this kind of sharing.  Maybe that was why you were concerned with the clean up?

Reply:
I think at this point I am confused about what you're concerned about.  There is nothing wrong with the lifecycle that I can see, but I may be missing something.

Reply:
I think after the dispose-disconnect work, I find it easier to reason about code and be confident that nothing spurious will be firing and causing troubles post-destroy if change handlers are explicitly put down.

At the moment, if the `DataViewItemModel` is still alive (due to reference cycle, say) even after the GUI is close down, its change handler will still be alive and respond to changes on the data model. Maybe that's okay. It doesn't make me feel all that comfortable the change handlers could still be alive.

Also I thought that perhaps the behaviour of `on_trait_change` with sharing objects might have affected the design, and the fact that those concerns are mitigated by `observe` might allow a different design that is more obviously correct.

Essentially, the following will work with `observe` but not with `on_trait_change`:

```
class MyDataModel(AbstractDataModel):
    ...

@provides(IDataViewWidget)
class MyDataView(MDataViewWidget, Widget):

    # this is a shared item model
    item_model = Instance(MyItemModel)

    def _create_control(self, parent):
        control = QTreeView(parent)
        control.setModel(self.item_model)
        self.item_model.start(self.data_model)

    def destroy(self):
        self.item_model.stop(self.data_model)
        # ... the rest of the destroy logic


class MyItemModel(QAbstractItemModel):

    def start(self, data_model):
        data_model.observe(self.on_values_changed, "values_changed", dispatch="ui")
        # ...

    def stop(self, data_model):
        data_model.observe(self.on_values_changed, "values_changed", dispatch="ui", remove=True)
```

Even if you have two views that use the same object for `item_model`, that's okay. Because `observe` is not idempotent: it keeps a reference count (see this [section](https://docs.enthought.com/traits/traits_user_manual/notification.html#hastraits-observe-is-not-idempotent)), to solve the exact problem when an object is shared.  So when one view calls `MyItemModel.stop` to disconnect the change handler, it would affect the second view.  Whereas `on_trait_change` would cause problem because it does not keep counts (which causes this [problem](https://github.com/enthought/traits/issues/237)).

Just wanted to point this out in case it might make you want to consider a different route that wasn't possible with on_trait_change.

Reply:
> So when one view calls `MyItemModel.stop` to disconnect the change handler, it would affect the second view.

Edited the comment because I missed the crucial "NOT" (typical Kit). It should read "it would NOT affect the second view".

Reply:
Just to clarify: It is okay if you prefer not to change the setter here. I was just trying to propose an alternative for considerations.

Reply:
Ok.  That's a bit clearer what you're worried about.

There is a really important difference in the `DataViewItemModel` case as compared to the places where you have been considering the `dispose` stuff: all of the examples where we have a `dispose` method, we are dealing with a Traits _wrapper_ around a Qt object; `DataViewItemModel` is not a wrapper, it _is_ a Qt object (actually, to be completely correct, it is a Qt object wrapping a Traits object).  This makes the lifecycle concerns completely different (and much simpler!).

Each `DataViewItemModel` holds a reference to an underlying `AbstractDataModel` instance, and it hooks up listeners (exactly one per trait, so idempotency is irrelevant) to the key change events.  Because these are methods on the class, they go through the method handling mechanisms of Traits observers, so when the `DataViewItemModel` gets GC'd then the handlers can no longer fire.

The one thing we need to be careful about is what to do if the underlying TraitsUI model changes: in this case we need to be careful to disconnect the dynamic observers from the old model, and connect new dynamic listeners to the old model, which is what this code is doing.

But it is certainly not any business of the widget to worry about hooking and unhooking observers on the `DataViewItemModel`.  It is its own thing and can potentially exist completely independently:
```
my_fancy_q_item_view = FancyView()
q_model = DataViewItemModel(model=traits_model)
my_fancy_q_view.setModel(q_model)
...
q_model.model = a_different_traits_model
...
```
We'll likely hash out exactly what use-cases we need to support when fleshing out the toolkit side of things.

Reply:
OK, I have come up with a potential case where we might get an exception, although not in any of the code here: if the `DataViewItemModel` has a _parent_ (which we don't do in any of our current code), and we delete the parent, then we can end up with a Python-side object where the underlying qt object has been destroyed, and then we will likely get an exception if a Trait handler fires.